contracts/contracts/Multisig.cairo

```diff
@@ -470,52 +426,131 @@ func execute_transaction{
     return (response_len=response.retdata_size, response=response.retdata)
 end
 
+# Sets number of required confirmations. The only way this can be invoked
+# is via a recursive call from execute_transaction -> set_confirmations_required.
+@external
+func set_confirmations_required{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    confirmations_required : felt
+):
+    require_multisig()
+
+    let (owners_len) = _owners_len.read()
+    _validate_and_set_confirmations_required(confirmations_required, owners_len)
+
+    return ()
+end
+
+# Sets the owners field on the multisig. The only way this can be invoked
+# is via a recursive call from execute_transaction -> set_owners.
+# Number of required confirmations is decreased in case its larger than
+# number of new owners.
+@external
+func set_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*
+):
+    require_multisig()
+    _set_owners(owners_len, owners)
+
+    return ()
+end
+
+# Set new owners and number of required confirmations.
+# Can be called only via recursively from
+# execute_transaction -> set_owners_and_confirmations_required
+@external
+func set_owners_and_confirmations_required{
+    syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr
+}(owners_len : felt, owners : felt*, confirmations_required : felt):
+    set_owners(owners_len, owners)
+    set_confirmations_required(confirmations_required)
+
+    return ()
+end
+
 #
 # Storage Helpers
 #
 
-func _set_owners{
-        syscall_ptr : felt*,
-        pedersen_ptr : HashBuiltin*,
-        range_check_ptr,
-    }(
-        owners_index : felt,
-        owners_len : felt,
-        owners : felt*,
-    ):
+func _get_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_index : felt, owners_len : felt, owners : felt*
+):
+    if owners_index == owners_len:
+        return ()
+    end
+
+    let (owner) = _owners.read(index=owners_index)
+    assert owners[owners_index] = owner
+
+    _get_owners(owners_index=owners_index + 1, owners_len=owners_len, owners=owners)
+    return ()
+end
+
+func _set_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*
+):
+    alloc_locals
+    require_unique_owners(owners_len, owners)
+
+    # Clean previous owners
+    let (old_owners_len) = _owners_len.read()
+    _clean_owners_range(0, old_owners_len)
+
+    let (local confirmations_required) = _confirmations_required.read()
+    # checks if owners_len < confirmations_required
+    let (lt) = is_le(owners_len, confirmations_required - 1)
+    if lt == TRUE:
+        # Lower number of confirmation
+        _confirmations_required.write(owners_len)
+        ConfirmationsSet.emit(owners_len)
```

Comment:
Additionally, the way the code is currently structured it will produce 2x `ConfirmationsSet` events when calling `set_owners_and_confirmations_required` fn with fewer owners than the current threshold.

This might introduce bugs where a user might expect 2x events (owners + threshold set), but it will actually get 3x events.

Let's restructure code like this:

```cairo
@constructor
func constructor{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
    owners_len : felt, owners : felt*, confirmations_required : felt
):
    _set_owners_and_confirmations_required(owners_len, owners, confirmations_required)

    return ()
end 

@external
func set_owners_and_confirmations_required{
    syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr
}(owners_len : felt, owners : felt*, confirmations_required : felt):
    require_multisig()
    _set_owners_and_confirmations_required(owners_len, owners, confirmations_required) 
}

func _set_owners_and_confirmations_required{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
    owners_len : felt, owners : felt*, confirmations_required : felt): 
    _set_owners(owners_len, owners)
    _validate_and_set_confirmations_required(confirmations_required, owners_len)
end

@external
func set_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
    owners_len : felt, owners : felt*
):
    require_multisig()
    _set_owners(owners_len, owners) // remove threshold check/set from here
    // TODO: check if new owners length less than current threshold, and if true call set_confirmations_required(owners_len)

    return ()
end
```

Reply:
Consider resolved (partly for now, but we can resolve this comment)!